File: orig_app/internal/handler/blob/send_file_upload.go
package blob

import (
	"fmt"
	"mime"
	"net/http"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/model"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
)

// SendFileUpload handles sending file uploads to clients
type SendFileUpload struct {
	workhorseClient *service.WorkhorseClient
	cdnService      *service.CDNService
}

// NewSendFileUpload creates a new instance of SendFileUpload
func NewSendFileUpload(
	workhorseClient *service.WorkhorseClient,
	cdnService *service.CDNService,
) *SendFileUpload {
	return &SendFileUpload{
		workhorseClient: workhorseClient,
		cdnService:      cdnService,
	}
}

// SendUpload sends a file upload to the client
func (s *SendFileUpload) SendUpload(
	c *gin.Context,
	fileUpload *model.FileUpload,
	sendParams map[string]interface{},
	redirectParams map[string]interface{},
	attachment string,
	proxy bool,
	disposition string,
) {
	// Default disposition to "attachment" if not provided
	if disposition == "" {
		disposition = "attachment"
	}

	// Get content type for the attachment
	contentType := s.contentTypeFor(attachment)

	// Handle attachment-specific parameters
	if attachment != "" {
		// Format the content disposition
		responseDisposition := fmt.Sprintf("%s; filename=\"%s\"", disposition, attachment)

		// Set response headers for cloud storage
		if redirectParams == nil {
			redirectParams = make(map[string]interface{})
		}
		if redirectParams["query"] == nil {
			redirectParams["query"] = make(map[string]string)
		}
		queryParams := redirectParams["query"].(map[string]string)
		queryParams["response-content-disposition"] = responseDisposition
		queryParams["response-content-type"] = contentType

		// Handle JavaScript files to avoid cross-origin protection
		if filepath.Ext(attachment) == ".js" {
			if sendParams == nil {
				sendParams = make(map[string]interface{})
			}
			sendParams["content_type"] = "text/plain"
		}

		// Set filename and disposition in send params
		if sendParams == nil {
			sendParams = make(map[string]interface{})
		}
		sendParams["filename"] = attachment
		sendParams["disposition"] = disposition
	}

	// Handle different types of file uploads
	if s.isImageScalingRequest(fileUpload, c) {
		// Handle image scaling request
		location := fileUpload.Path
		if !fileUpload.IsFileStorage() {
			location = fileUpload.URL
		}

		width := c.Query("width")
		widthInt := 0
		if width != "" {
			fmt.Sscanf(width, "%d", &widthInt)
		}

		// Send scaled image through workhorse
		headers := s.workhorseClient.SendScaledImage(location, widthInt, contentType)
		for key, value := range headers {
			c.Header(key, value)
		}
		c.Status(http.StatusOK)
	} else if fileUpload.IsFileStorage() {
		// Send file directly
		c.File(fileUpload.Path)
	} else if fileUpload.IsProxyDownloadEnabled() || proxy {
		// Send URL through workhorse
		headers := s.workhorseClient.SendURL(fileUpload.URL, redirectParams)
		for key, value := range headers {
			c.Header(key, value)
		}
		c.Status(http.StatusOK)
	} else {
		// Redirect to CDN URL
		clientIP := c.ClientIP()
		fileURL := s.cdnService.GetFileURL(fileUpload, clientIP, redirectParams)
		c.Redirect(http.StatusFound, fileURL)
	}
}

// ContentTypeFor returns the content type for an attachment
func (s *SendFileUpload) ContentTypeFor(attachment string) string {
	if attachment == "" {
		return ""
	}
	return s.guessContentType(attachment)
}

// GuessContentType guesses the content type for a filename
func (s *SendFileUpload) GuessContentType(filename string) string {
	contentType := mime.TypeByExtension(filepath.Ext(filename))
	if contentType != "" {
		return contentType
	}
	return "application/octet-stream"
}

// IsImageScalingRequest checks if the request is for image scaling
func (s *SendFileUpload) IsImageScalingRequest(fileUpload *model.FileUpload, c *gin.Context) bool {
	return s.isAvatarSafeForScaling(fileUpload, c) || s.isPwaIconSafeForScaling(fileUpload, c)
}

// IsPwaIconSafeForScaling checks if a PWA icon is safe for scaling
func (s *SendFileUpload) IsPwaIconSafeForScaling(fileUpload *model.FileUpload, c *gin.Context) bool {
	width := c.Query("width")
	widthInt := 0
	if width != "" {
		fmt.Sscanf(width, "%d", &widthInt)
	}

	return fileUpload.IsImageSafeForScaling() &&
		s.isMountedAsPwaIcon(fileUpload) &&
		s.isValidImageScalingWidth(widthInt, []int{192, 512}) // Example allowed widths
}

// IsAvatarSafeForScaling checks if an avatar is safe for scaling
func (s *SendFileUpload) IsAvatarSafeForScaling(fileUpload *model.FileUpload, c *gin.Context) bool {
	width := c.Query("width")
	widthInt := 0
	if width != "" {
		fmt.Sscanf(width, "%d", &widthInt)
	}

	return fileUpload.IsImageSafeForScaling() &&
		s.isMountedAsAvatar(fileUpload) &&
		s.isValidImageScalingWidth(widthInt, []int{32, 64, 128}) // Example allowed widths
}

// IsMountedAsAvatar checks if a file upload is mounted as an avatar
func (s *SendFileUpload) IsMountedAsAvatar(fileUpload *model.FileUpload) bool {
	return fileUpload.MountedAs == "avatar"
}

// IsMountedAsPwaIcon checks if a file upload is mounted as a PWA icon
func (s *SendFileUpload) IsMountedAsPwaIcon(fileUpload *model.FileUpload) bool {
	return fileUpload.MountedAs == "pwa_icon"
}

// IsValidImageScalingWidth checks if the requested width is valid for scaling
func (s *SendFileUpload) IsValidImageScalingWidth(width int, allowedScalarWidths []int) bool {
	for _, allowedWidth := range allowedScalarWidths {
		if allowedWidth == width {
			return true
		}
	}
	return false
}

// Private helper methods
func (s *SendFileUpload) contentTypeFor(attachment string) string {
	if attachment == "" {
		return ""
	}
	return s.guessContentType(attachment)
}

func (s *SendFileUpload) guessContentType(filename string) string {
	contentType := mime.TypeByExtension(filepath.Ext(filename))
	if contentType != "" {
		return contentType
	}
	return "application/octet-stream"
}

func (s *SendFileUpload) isImageScalingRequest(fileUpload *model.FileUpload, c *gin.Context) bool {
	return s.isAvatarSafeForScaling(fileUpload, c) || s.isPwaIconSafeForScaling(fileUpload, c)
}

func (s *SendFileUpload) isPwaIconSafeForScaling(fileUpload *model.FileUpload, c *gin.Context) bool {
	width := c.Query("width")
	widthInt := 0
	if width != "" {
		fmt.Sscanf(width, "%d", &widthInt)
	}

	return fileUpload.IsImageSafeForScaling() &&
		s.isMountedAsPwaIcon(fileUpload) &&
		s.isValidImageScalingWidth(widthInt, []int{192, 512}) // Example allowed widths
}

func (s *SendFileUpload) isAvatarSafeForScaling(fileUpload *model.FileUpload, c *gin.Context) bool {
	width := c.Query("width")
	widthInt := 0
	if width != "" {
		fmt.Sscanf(width, "%d", &widthInt)
	}

	return fileUpload.IsImageSafeForScaling() &&
		s.isMountedAsAvatar(fileUpload) &&
		s.isValidImageScalingWidth(widthInt, []int{32, 64, 128}) // Example allowed widths
}

func (s *SendFileUpload) isMountedAsAvatar(fileUpload *model.FileUpload) bool {
	return fileUpload.MountedAs == "avatar"
}

func (s *SendFileUpload) isMountedAsPwaIcon(fileUpload *model.FileUpload) bool {
	return fileUpload.MountedAs == "pwa_icon"
}

func (s *SendFileUpload) isValidImageScalingWidth(width int, allowedScalarWidths []int) bool {
	for _, allowedWidth := range allowedScalarWidths {
		if allowedWidth == width {
			return true
		}
	}
	return false
}

File: src/uploads/send_file_upload.rs
use actix_web::{HttpResponse, Responder};
use std::collections::HashMap;
use std::path::Path;

pub trait SendFileUpload {
    fn send_upload(
        &self,
        file_uploader: Box<dyn crate::uploads::Uploader>,
        send_params: HashMap<String, String>,
        redirect_params: HashMap<String, String>,
        attachment: Option<String>,
        proxy: bool,
        disposition: String,
    ) -> impl Responder;

    fn content_type_for(&self, attachment: Option<&str>) -> String;
    fn guess_content_type(&self, filename: &str) -> String;
    fn image_scaling_request(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool;
    fn pwa_icon_safe_for_scaling(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool;
    fn avatar_safe_for_scaling(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool;
    fn mounted_as_avatar(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool;
    fn mounted_as_pwa_icon(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool;
    fn valid_image_scaling_width(&self, allowed_scalar_widths: &[i32]) -> bool;
    fn build_file_url(
        &self,
        file_uploader: &dyn crate::uploads::Uploader,
        redirect_params: HashMap<String, String>,
    ) -> String;
}

impl<T> SendFileUpload for T
where
    T: crate::uploads::UploadsActionsHandler,
{
    fn send_upload(
        &self,
        file_uploader: Box<dyn crate::uploads::Uploader>,
        send_params: HashMap<String, String>,
        redirect_params: HashMap<String, String>,
        attachment: Option<String>,
        proxy: bool,
        disposition: String,
    ) -> impl Responder {
        let content_type = self.content_type_for(attachment.as_deref());

        let mut response = HttpResponse::Ok();

        if let Some(attachment) = attachment {
            let content_disposition = format!("{}; filename=\"{}\"", disposition, attachment);
            response.append_header(("Content-Disposition", content_disposition));

            // Handle JS files specially
            if attachment.ends_with(".js") {
                response.append_header(("Content-Type", "text/plain"));
            } else {
                response.append_header(("Content-Type", content_type));
            }
        }

        if self.image_scaling_request(&file_uploader) {
            // Handle image scaling
            let location = if file_uploader.file_storage() {
                file_uploader.path()
            } else {
                file_uploader.url(HashMap::new())
            };

            // In a real implementation, this would set headers for scaled image
            response.append_header(("X-Scaled-Image", location));
            response
        } else if file_uploader.file_storage() {
            // Send file directly
            response.append_header(("Content-Type", content_type));
            response.append_header((
                "Content-Disposition",
                format!(
                    "{}; filename=\"{}\"",
                    disposition,
                    attachment.unwrap_or_default()
                ),
            ));
            response.body("File content would be here")
        } else if file_uploader.proxy_download_enabled() || proxy {
            // Proxy download
            let url = file_uploader.url(redirect_params);
            response.append_header(("X-Send-File", url));
            response
        } else {
            // Redirect to file URL
            let file_url = self.build_file_url(&file_uploader, redirect_params);
            HttpResponse::Found()
                .append_header(("Location", file_url))
                .finish()
        }
    }

    fn content_type_for(&self, attachment: Option<&str>) -> String {
        match attachment {
            Some(filename) => self.guess_content_type(filename),
            None => "application/octet-stream".to_string(),
        }
    }

    fn guess_content_type(&self, filename: &str) -> String {
        // In a real implementation, this would use a MIME type library
        match Path::new(filename).extension().and_then(|ext| ext.to_str()) {
            Some("jpg") | Some("jpeg") => "image/jpeg".to_string(),
            Some("png") => "image/png".to_string(),
            Some("gif") => "image/gif".to_string(),
            Some("pdf") => "application/pdf".to_string(),
            Some("js") => "text/javascript".to_string(),
            Some("css") => "text/css".to_string(),
            Some("html") => "text/html".to_string(),
            Some("txt") => "text/plain".to_string(),
            _ => "application/octet-stream".to_string(),
        }
    }

    fn image_scaling_request(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool {
        self.avatar_safe_for_scaling(file_uploader) || self.pwa_icon_safe_for_scaling(file_uploader)
    }

    fn pwa_icon_safe_for_scaling(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool {
        file_uploader.image_safe_for_scaling()
            && self.mounted_as_pwa_icon(file_uploader)
            && self.valid_image_scaling_width(&[16, 32, 64, 128, 192, 512])
    }

    fn avatar_safe_for_scaling(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool {
        file_uploader.image_safe_for_scaling()
            && self.mounted_as_avatar(file_uploader)
            && self.valid_image_scaling_width(&[32, 64, 128])
    }

    fn mounted_as_avatar(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool {
        file_uploader
            .mounted_as()
            .map_or(false, |mounted_as| mounted_as == "avatar")
    }

    fn mounted_as_pwa_icon(&self, file_uploader: &dyn crate::uploads::Uploader) -> bool {
        file_uploader
            .mounted_as()
            .map_or(false, |mounted_as| mounted_as == "pwa_icon")
    }

    fn valid_image_scaling_width(&self, allowed_scalar_widths: &[i32]) -> bool {
        // In a real implementation, this would check if the width is in the allowed list
        allowed_scalar_widths.contains(&32)
    }

    fn build_file_url(
        &self,
        file_uploader: &dyn crate::uploads::Uploader,
        redirect_params: HashMap<String, String>,
    ) -> String {
        // In a real implementation, this would build a CDN URL
        file_uploader.url(redirect_params)
    }
}
